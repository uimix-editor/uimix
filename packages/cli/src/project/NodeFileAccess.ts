import * as fs from "fs";
import * as path from "path";
import { FileAccess, Stats } from "./FileAccess";
import chokidar from "chokidar";
import { mkdirp } from "mkdirp";
import { globby } from "globby";

export class NodeFileAccess implements FileAccess {
  watch(cwd: string, patterns: string[], onChange: () => void): () => void {
    const watcher = chokidar.watch(`**/{${patterns.join(",")},}`, {
      cwd,
      ignored: ["**/node_modules/**", "**/.git/**"],
    });
    watcher.on("change", onChange);
    watcher.on("add", onChange);
    watcher.on("unlink", onChange);
    return () => watcher.close();
  }

  async glob(cwd: string, patterns: string[]): Promise<string[]> {
    return (
      await globby(`**/{${patterns.join(",")},}`, {
        cwd,
        ignore: ["**/node_modules/**", "**/.git/**"],
      })
    ).map((filePath) => path.resolve(cwd, filePath));
  }

  async stat(filePath: string): Promise<Stats | undefined> {
    try {
      const stat = await fs.promises.stat(filePath);
      return {
        type: stat.isDirectory() ? "directory" : "file",
      };
    } catch (e) {
      return undefined;
    }
  }

  async writeFile(filePath: string, data: Buffer): Promise<void> {
    await mkdirp(path.dirname(filePath));
    await fs.promises.writeFile(filePath, data);
  }

  async readFile(filePath: string): Promise<Buffer> {
    return await fs.promises.readFile(filePath);
  }

  async remove(filePath: string): Promise<void> {
    await fs.promises.rm(filePath);
  }
}
