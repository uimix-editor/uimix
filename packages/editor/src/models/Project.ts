/* eslint-disable @typescript-eslint/no-explicit-any */

import * as Y from "yjs";
import { posix as path } from "path-browserify";
import { getOrCreate } from "../state/Collection";
import { SelectableMap } from "./Selectable";
import { Node, NodeMap } from "./Node";
import { computed, makeObservable, ObservableMap } from "mobx";
import { ProjectJSON } from "@uimix/node-data";
import { toProjectJSON } from "./toProjectJSON";
import { ImageManager } from "./ImageManager";
import { Component } from "./Component";
import { ObservableYMap } from "../utils/ObservableYMap";

export interface PageHierarchyFolderEntry {
  type: "directory";
  path: string;
  name: string;
  children: PageHierarchyEntry[];
}

export interface PageHierarchyPageEntry {
  type: "file";
  path: string;
  name: string;
  page: Node;
}

export type PageHierarchyEntry =
  | PageHierarchyFolderEntry
  | PageHierarchyPageEntry;

class Pages {
  constructor(project: Project) {
    this.project = project;
    this.node = project.node;
    makeObservable(this);
  }

  node: Node;
  readonly project: Project;

  @computed get count(): number {
    return this.node.childCount;
  }

  get all(): Node[] {
    return this.node.children;
  }

  create(filePath: string): Node {
    const node = this.project.nodes.create("page");
    node.name = filePath;
    return node;
  }

  toHierarchy(): PageHierarchyFolderEntry {
    const root: PageHierarchyFolderEntry = {
      type: "directory",
      name: "",
      path: "",
      children: [],
    };
    const parents = new Map<string, PageHierarchyFolderEntry>();
    parents.set("", root);

    const mkdirp = (segments: string[]): PageHierarchyFolderEntry => {
      if (segments.length === 0) {
        return root;
      }

      const existing = parents.get(segments.join("/"));
      if (existing) {
        return existing;
      }

      const parent = mkdirp(segments.slice(0, -1));
      const dir: PageHierarchyFolderEntry = {
        type: "directory",
        name: segments[segments.length - 1],
        path: segments.join("/"),
        children: [],
      };
      parent.children.push(dir);
      parents.set(segments.join("/"), dir);
      return dir;
    };

    const pages = Array.from(this.all);
    pages.sort((a, b) => a.name.localeCompare(b.name));

    for (const page of pages) {
      const segments = page.name.split(path.sep);
      const parent = mkdirp(segments.slice(0, -1));

      const item: PageHierarchyPageEntry = {
        type: "file",
        name: segments[segments.length - 1],
        path: page.name,
        page,
      };
      parent.children.push(item);
    }

    return root;
  }

  affectedPagesForPath(path: string): Node[] {
    return this.all.filter(
      (page) => page.name === path || page.name.startsWith(path + "/")
    );
  }

  deletePageOrPageFolder(path: string) {
    const pagesToDelete = this.affectedPagesForPath(path);

    for (const page of pagesToDelete) {
      page.remove();
      // TODO: delete dangling nodes?
      //this.project.nodes.remove(doc.root);
    }
  }

  renamePageOrPageFolder(path: string, newPath: string) {
    if (path === newPath) {
      return;
    }

    const pagesToDelete = this.affectedPagesForPath(path);

    for (const page of pagesToDelete) {
      page.name = newPath + page.name.slice(path.length);
    }
  }
}

export class Project {
  constructor(data: Y.Map<any>) {
    this.data = ObservableYMap.get(data);
    this.nodes = new NodeMap(this);
    this.selectables = new SelectableMap(this);
    this.node = this.nodes.create("project", "project");
    this.pages = new Pages(this);
  }

  readonly data: ObservableYMap<any>;
  readonly nodes: NodeMap;
  readonly selectables: SelectableMap;
  readonly node: Node;
  readonly pages: Pages;
  readonly imageManager = new ImageManager();

  toJSON(): ProjectJSON {
    return toProjectJSON(this.node.children.map((c) => c.selectable));
  }

  loadJSON(json: ProjectJSON) {
    this.node.clear();

    for (const [id, nodeJSON] of Object.entries(json.nodes)) {
      const node = this.nodes.getOrCreate(nodeJSON.type, id);
      node.loadJSON(nodeJSON);
      if (node.type === "page") {
        this.node.append([node]);
      }
    }
    for (const [id, style] of Object.entries(json.styles)) {
      const selectable = this.selectables.get(id.split(":"));
      selectable.selfStyle.loadJSON(style);
    }
  }

  get components(): Component[] {
    const components: Component[] = [];

    for (const page of this.pages.all) {
      for (const node of page.children) {
        const component = Component.from(node);
        if (component) {
          components.push(component);
        }
      }
    }

    return components;
  }
}
