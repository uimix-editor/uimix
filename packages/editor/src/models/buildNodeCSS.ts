import { NodeType, StackDirection, StyleJSON } from "@uimix/node-data";

export function getLayoutType(
  style: StyleJSON
): StackDirection | "grid" | undefined {
  if (style.layout === "flex") {
    return style.flexDirection;
  }
  if (style.layout === "grid") {
    return "grid";
  }
}

export function buildNodeCSS(
  nodeType: NodeType,
  style: StyleJSON,
  parentLayout?: StackDirection | "grid"
): React.CSSProperties {
  if (nodeType === "component") {
    return {};
  }

  const cssStyle: React.CSSProperties = {};

  const cssPosition = parentLayout && !style.absolute ? "relative" : "absolute";
  cssStyle.position = cssPosition;
  if (cssPosition === "absolute") {
    const position = style.position;
    if ("start" in position.x) {
      cssStyle.left = `${position.x.start}px`;
    }
    if ("end" in position.x) {
      cssStyle.right = `${position.x.end}px`;
    }

    if ("start" in position.y) {
      cssStyle.top = `${position.y.start}px`;
    }
    if ("end" in position.y) {
      cssStyle.bottom = `${position.y.end}px`;
    }
  } else {
    cssStyle.marginTop = `${style.marginTop}px`;
    cssStyle.marginRight = `${style.marginRight}px`;
    cssStyle.marginBottom = `${style.marginBottom}px`;
    cssStyle.marginLeft = `${style.marginLeft}px`;
  }

  // TODO: unset width/height when both left/right or top/bottom are set

  const width = style.width;
  if (width.type === "fixed") {
    cssStyle.width = `${width.value}px`;
  } else if (width.type === "hug") {
    cssStyle.width = "max-content";
  } else {
    if (parentLayout === "x") {
      cssStyle.flex = 1;
    } else if (parentLayout) {
      cssStyle.width = `calc(100% - ${style.marginLeft + style.marginRight}px)`;
    } else {
      cssStyle.width = "auto";
    }
    cssStyle.minWidth = width.min !== undefined ? `${width.min}px` : undefined;
    cssStyle.maxWidth = width.max !== undefined ? `${width.max}px` : undefined;
  }

  const height = style.height;
  if (height.type === "fixed") {
    cssStyle.height = `${height.value}px`;
  } else if (height.type === "hug") {
    cssStyle.height = "max-content";
  } else {
    if (parentLayout === "y") {
      cssStyle.flex = 1;
    } else if (parentLayout) {
      cssStyle.height = `calc(100% - ${
        style.marginTop + style.marginBottom
      }px)`;
    } else {
      cssStyle.height = "auto";
    }
    cssStyle.minHeight =
      height.min !== undefined ? `${height.min}px` : undefined;
    cssStyle.maxHeight =
      height.max !== undefined ? `${height.max}px` : undefined;
  }

  cssStyle.opacity = style.opacity;
  cssStyle.overflow = style.overflowHidden ? "hidden" : "visible";

  if (nodeType === "frame") {
    const layout = style.layout;
    if (layout === "flex") {
      cssStyle.display = "flex";
      cssStyle.flexDirection = style.flexDirection === "x" ? "row" : "column";
      cssStyle.alignItems = (() => {
        switch (style.flexAlign) {
          case "start":
            return "flex-start";
          case "center":
            return "center";
          case "end":
            return "flex-end";
        }
      })();
      cssStyle.justifyContent = (() => {
        switch (style.flexJustify) {
          case "start":
            return "flex-start";
          case "center":
            return "center";
          case "end":
            return "flex-end";
          case "spaceBetween":
            return "space-between";
        }
      })();
      cssStyle.rowGap = `${style.rowGap}px`;
      cssStyle.columnGap = `${style.columnGap}px`;
    } else if (layout === "grid") {
      cssStyle.display = "grid";
      const { gridRowCount, gridColumnCount } = style;
      if (gridRowCount !== null) {
        cssStyle.gridTemplateRows = `repeat(${gridRowCount}, 1fr)`;
      }
      if (gridColumnCount !== null) {
        cssStyle.gridTemplateColumns = `repeat(${gridColumnCount}, 1fr)`;
      }
      cssStyle.rowGap = `${style.rowGap}px`;
      cssStyle.columnGap = `${style.columnGap}px`;
    } else {
      cssStyle.display = "block";
    }
    cssStyle.paddingLeft = `${style.paddingLeft}px`;
    cssStyle.paddingRight = `${style.paddingRight}px`;
    cssStyle.paddingTop = `${style.paddingTop}px`;
    cssStyle.paddingBottom = `${style.paddingBottom}px`;
  }

  if (nodeType === "frame" || nodeType === "image" || nodeType === "svg") {
    const fills = style.fills;
    cssStyle.background = fills.length ? fills[0].color : "transparent";
    cssStyle.borderStyle = "solid";
    cssStyle.borderColor = style.border?.color ?? "transparent";
    cssStyle.borderTopWidth = `${style.borderTopWidth}px`;
    cssStyle.borderRightWidth = `${style.borderRightWidth}px`;
    cssStyle.borderBottomWidth = `${style.borderBottomWidth}px`;
    cssStyle.borderLeftWidth = `${style.borderLeftWidth}px`;

    cssStyle.borderTopLeftRadius = `${style.topLeftRadius}px`;
    cssStyle.borderTopRightRadius = `${style.topRightRadius}px`;
    cssStyle.borderBottomRightRadius = `${style.bottomRightRadius}px`;
    cssStyle.borderBottomLeftRadius = `${style.bottomLeftRadius}px`;

    const shadows = style.shadows;
    if (shadows.length === 0) {
      cssStyle.boxShadow = "none";
    } else {
      cssStyle.boxShadow = shadows
        .map((shadow) => {
          const x = `${shadow.x}px`;
          const y = `${shadow.y}px`;
          const blur = `${shadow.blur}px`;
          const spread = `${shadow.spread}px`;
          const color = shadow.color;
          return `${x} ${y} ${blur} ${spread} ${color}`;
        })
        .join(", ");
    }
  }

  if (nodeType === "text") {
    cssStyle.whiteSpace = "break-spaces";
    cssStyle.display = "flex";
    cssStyle.flexDirection = "column";
    const fills = style.fills;
    cssStyle.color = fills.length ? fills[0].color : "transparent";
    cssStyle.fontFamily = style.fontFamily;
    cssStyle.fontSize = `${style.fontSize}px`;
    cssStyle.fontWeight = style.fontWeight;
    const lineHeight = style.lineHeight;
    cssStyle.lineHeight =
      lineHeight === null
        ? "normal"
        : typeof lineHeight === "number"
        ? `${lineHeight}px`
        : lineHeight.join("");
    const letterSpacing = style.letterSpacing;
    cssStyle.letterSpacing =
      typeof letterSpacing === "number"
        ? `${letterSpacing}px`
        : `${letterSpacing[0] / 100}em`;
    cssStyle.textAlign = style.textHorizontalAlign;
    cssStyle.justifyContent = (() => {
      switch (style.textVerticalAlign) {
        case "start":
          return "flex-start";
        case "center":
          return "center";
        case "end":
          return "flex-end";
      }
    })();
  }

  if (style.hidden) {
    cssStyle.display = "none";
  }

  return cssStyle;
}
