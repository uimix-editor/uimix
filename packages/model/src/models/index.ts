export * from "./ColorRef";
export * from "./ColorToken";
export * from "./Component";
export * from "./ImageManager";
export * from "./Node";
export * from "./ObjectData";
export * from "./Page";
export * from "./PageList";
export * from "./Project";
export * from "../collaborative/ProjectData";
export * from "./Selectable";
export * from "./Style";
