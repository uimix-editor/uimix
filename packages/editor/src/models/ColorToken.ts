import { Color } from "../utils/Color";
import * as Y from "yjs";
import { ObservableYMap } from "../utils/ObservableYMap";
import { Project } from "./Project";
import { getOrCreate } from "../state/Collection";
import { generateID } from "../utils/ID";

export class ColorToken {
  constructor(project: Project, id: string) {
    this.id = id;
    this.project = project;
  }

  readonly id: string;
  readonly project: Project;

  get data(): ObservableYMap<unknown> | undefined {
    return ObservableYMap.get(this.project.colorTokens.data.get(this.id));
  }

  get dataForWrite(): ObservableYMap<unknown> {
    return ObservableYMap.get(
      getOrCreate(this.project.colorTokens.data, this.id, () => new Y.Map())
    );
  }

  get name(): string | undefined {
    return this.data?.get("name") as string | undefined;
  }

  set name(value: string | undefined) {
    if (value === undefined) {
      this.data?.delete("name");
    } else {
      this.dataForWrite.set("name", value);
    }
  }

  get value(): Color | undefined {
    const hex = this.data?.get("value") as string | undefined;
    if (!hex) {
      return undefined;
    }
    return Color.from(hex);
  }

  set value(value: Color | undefined) {
    if (value === undefined) {
      this.data?.delete("value");
    } else {
      this.dataForWrite.set("value", value.toHex());
    }
  }

  get index(): number {
    return (this.data?.get("index") as number | undefined) ?? 0;
  }

  set index(value: number) {
    this.dataForWrite.set("index", value);
  }
}

export class ColorTokenList {
  constructor(project: Project) {
    this.project = project;
  }

  readonly project: Project;

  get data(): ObservableYMap<Y.Map<unknown>> {
    return ObservableYMap.get(this.project.doc.getMap("colors"));
  }

  get(id: string): ColorToken | undefined {
    if (!this.data.has(id)) {
      return undefined;
    }
    return new ColorToken(this.project, id);
  }

  get all(): ColorToken[] {
    return [...this.data.keys()]
      .map((id) => new ColorToken(this.project, id))
      .sort((a, b) => a.index - b.index);
  }

  add(): ColorToken {
    const id = generateID();

    const token = new ColorToken(this.project, id);
    token.index = this.all.length;
    token.name = "New Color";
    return token;
  }
}
