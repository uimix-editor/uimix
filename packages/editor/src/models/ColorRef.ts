import { Color } from "../utils/Color";
import { ColorToken } from "./ColorToken";
import { Project } from "./Project";
import { Color as ColorJSON } from "@uimix/node-data";

export class ColorRef {
  constructor(value: Color | ColorToken) {
    this.value = value;
  }

  static fromJSON(project: Project, json: ColorJSON): ColorRef | undefined {
    if (typeof json === "string") {
      const color = Color.from(json);
      return color && new ColorRef(color);
    }
    const token = project.colorTokens.get(json.id);
    return token && new ColorRef(token);
  }

  toJSON(): ColorJSON {
    if (this.value instanceof Color) {
      return this.value.toHex();
    } else {
      return {
        type: "token",
        id: this.value.id,
      };
    }
  }

  get color(): Color {
    if (this.value instanceof Color) {
      return this.value;
    } else {
      return this.value.value ?? Color.black;
    }
  }

  readonly value: Color | ColorToken;
}
