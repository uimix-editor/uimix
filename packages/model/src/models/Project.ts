import * as Y from "yjs";
import { Selectable, SelectableMap } from "./Selectable";
import { Node, NodeMap } from "./Node";
import * as Data from "../data/v1";
import { ImageManager } from "./ImageManager";
import { Component } from "./Component";
import { ObservableYArray } from "@uimix/foundation/src/utils/ObservableYArray";
import { ColorTokenMap } from "./ColorToken";
import { PageList } from "./PageList";
import { ProjectData } from "../collaborative/ProjectData";
import { CodeAssets } from "./CodeAssets";
import { makeObservable, observable } from "mobx";
import { Page } from "./Page";

export class Project {
  constructor() {
    this.nodes = new NodeMap(this);
    this.selectables = new SelectableMap(this);
    this.colorTokens = new ColorTokenMap(this);
    this.node = this.nodes.create("project", "project");
    this.pages = new PageList(this);
    makeObservable(this);
  }

  readonly data = new ProjectData();
  readonly nodes: NodeMap;
  readonly selectables: SelectableMap;
  readonly colorTokens: ColorTokenMap;
  readonly node: Node;
  readonly pages: PageList;
  readonly imageManager = new ImageManager(this);

  get doc(): Y.Doc {
    return this.data.doc;
  }

  createUndoManager(): Y.UndoManager {
    return new Y.UndoManager([
      this.data.nodes,
      this.data.styles,
      this.data.selection,
      this.data.componentURLs,
      this.data.colors,
    ]);
  }

  clear(): void {
    this.data.clear();
    this.nodes.create("project", "project");
  }

  toJSON(): Data.Project {
    return this.data.toJSON();
  }

  loadJSON(json: Data.Project) {
    this.data.loadJSON(json);
  }

  get components(): Component[] {
    return this.pages.all.flatMap((page) => page.components);
  }

  get componentURLs(): ObservableYArray<string> {
    return ObservableYArray.get(this.data.componentURLs);
  }

  @observable localCodeAssets: CodeAssets | undefined = undefined;

  // Selection

  clearSelection() {
    this.selectables.selectionData.clear();
  }

  replaceSelection(selectables: Iterable<Selectable>) {
    this.clearSelection();
    for (const selectable of selectables) {
      selectable.select();
    }
  }

  // Get by ID

  nodeForID(id: string): Node | undefined {
    return this.nodes.get(id);
  }

  pageForID(id: string): Page | undefined {
    const node = this.nodeForID(id);
    if (node) {
      return Page.from(node);
    }
  }

  componentForID(id: string): Component | undefined {
    const node = this.nodeForID(id);
    if (node) {
      return Component.from(node);
    }
  }
}
