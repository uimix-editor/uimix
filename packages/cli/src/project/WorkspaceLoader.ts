import {
  Image,
  ProjectManifestJSON,
  ImageType,
} from "@uimix/model/src/data/v1";
import { usedImageHashesInProject } from "@uimix/model/src/data/util";
import { isEqual } from "lodash-es";
import { formatTypeScript } from "../format";
import { FileAccess } from "./FileAccess";
import * as path from "path";
import { Project } from "@uimix/model/src/models";
import {
  ProjectEmitter,
  ProjectLoader,
  PageNode,
  loadFromJSXFile,
  stringifyAsJSXFile,
} from "@uimix/model/src/file";
import { getURLSafeBase64Hash } from "@uimix/foundation/src/utils/Hash";
import sizeOf from "image-size";
import { dataUriToBuffer } from "data-uri-to-buffer";
import * as mime from "mime-types";

interface ProjectData {
  manifest: ProjectManifestJSON;
  project: Project;
  pages: Map<string, PageNode>;
  imagePaths: Map<string, string>; // hash to path
}

interface LoadProblem {
  filePath: string;
  error: unknown;
}

interface LoadResult {
  changed: boolean;
  problems: LoadProblem[];
}

// Important TODO: fix paths in Windows!!
export class WorkspaceLoader {
  static async load(fileAccess: FileAccess) {
    const loader = new WorkspaceLoader(fileAccess);
    const result = await loader.load();
    if (result.problems.length) {
      throw new Error(
        `Problems loading workspace:\n${result.problems
          .map(
            (problem) => `  ${problem.filePath}:\n    ${String(problem.error)}`
          )
          .join("\n")}`
      );
    }
    return loader;
  }

  constructor(fileAccess: FileAccess) {
    this.fileAccess = fileAccess;
  }

  readonly fileAccess: FileAccess;

  get rootPath(): string {
    return this.fileAccess.rootPath;
  }

  readonly filePattern = "**/*.uimix";
  readonly imagePatterns = ["**/*.png", "**/*.jpg", "**/*.jpeg"];
  readonly uimixProjectFile = "uimix.json";
  readonly projectBoundary = "package.json"; // TODO: other project boundaries
  projects = new Map<string, ProjectData>(); // project path -> project json

  get rootProject(): ProjectData {
    return this.getOrCreateProject(this.rootPath);
  }

  getOrCreateProject(projectPath: string): ProjectData {
    let project = this.projects.get(projectPath);
    if (!project) {
      project = {
        manifest: {},
        project: new Project(),
        pages: new Map(),
        imagePaths: new Map(),
      };
      this.projects.set(projectPath, project);
    }
    return project;
  }

  projectPathForFile(pagePath: string): string {
    return (
      [...this.projects.keys()]
        .sort((a, b) => b.length - a.length)
        .find((projectPath) => pagePath.startsWith(projectPath + path.sep)) ??
      this.rootPath
    );
  }

  async load(): Promise<LoadResult> {
    const filePaths = await this.fileAccess.glob(
      `{${this.filePattern},${this.imagePatterns.join(",")},**/${
        this.projectBoundary
      }}`
    );
    filePaths.sort();

    const filePathsForProject = new Map<string, string[]>([
      [this.rootPath, []],
    ]);

    for (const manifestPaths of filePaths.filter((filePath) =>
      filePath.endsWith(this.projectBoundary)
    )) {
      const parentPath = path.dirname(manifestPaths);
      filePathsForProject.set(parentPath, []);
    }

    for (const filePath of filePaths) {
      if (filePath.endsWith(this.projectBoundary)) {
        continue;
      }

      const projectPath = [...filePathsForProject.keys()]
        .sort((a, b) => b.length - a.length)
        .find((projectPath) => filePath.startsWith(projectPath + path.sep));
      if (!projectPath) {
        throw new Error("not supposed to happen");
      }
      filePathsForProject.get(projectPath)?.push(filePath);
    }

    let changed = false;
    const problems: LoadProblem[] = [];
    for (const [projectPath, filePaths] of filePathsForProject) {
      const result = await this.loadProject(projectPath, filePaths);
      changed = changed || result.changed;
      problems.push(...result.problems);
    }
    return {
      changed,
      problems,
    };
  }

  private async loadProject(
    projectPath: string,
    filePaths: string[]
  ): Promise<LoadResult> {
    const problems: LoadProblem[] = [];

    try {
      let manifest: ProjectManifestJSON = {};

      const manifestPath = path.join(projectPath, this.uimixProjectFile);
      if ((await this.fileAccess.stat(manifestPath))?.type === "file") {
        try {
          manifest = ProjectManifestJSON.parse(
            JSON.parse(
              (
                await this.fileAccess.readFile(
                  path.join(projectPath, this.uimixProjectFile)
                )
              ).toString()
            )
          );
        } catch (error) {
          problems.push({
            filePath: manifestPath,
            error,
          });
        }
      }

      const images = new Map<string, Image>();
      const pages = new Map<string, PageNode>();
      const imagePaths = new Map<string, string>();

      for (const filePath of filePaths) {
        // TODO: reload changed files only

        if (filePath.endsWith(".uimix")) {
          try {
            const pageText = (
              await this.fileAccess.readFile(filePath)
            ).toString();
            const pageNode = loadFromJSXFile(pageText);
            const pageName = path
              .relative(projectPath, filePath)
              .replace(/\.uimix$/, "");
            pages.set(pageName, pageNode);
          } catch (error) {
            problems.push({
              filePath,
              error,
            });
          }
        } else {
          try {
            // TODO: lookup specific directories only
            const imageData = await this.fileAccess.readFile(filePath);
            const hash = await getURLSafeBase64Hash(imageData);
            const mimeType = mime.lookup(filePath) || "image/png";
            const size = sizeOf(imageData);

            const image: Image = {
              width: size.width ?? 0,
              height: size.height ?? 0,
              type: mimeType as ImageType,
              url: `data:${mimeType};base64,${imageData.toString("base64")}`,
            };
            images.set(hash, image);
            imagePaths.set(hash, filePath);
          } catch (error) {
            problems.push({
              filePath,
              error,
            });
          }
        }
      }

      const existingProject = this.projects.get(projectPath);

      const loader = new ProjectLoader(existingProject?.project);
      loader.load(pages);
      for (const [key, image] of images) {
        loader.project.imageManager.images.set(key, image);
      }

      if (isEqual(pages, existingProject?.pages ?? new Map())) {
        return {
          changed: false,
          problems,
        };
      }

      this.projects.set(projectPath, {
        manifest,
        project: loader.project,
        pages,
        imagePaths,
      });

      return {
        changed: true,
        problems,
      };
    } catch (error) {
      problems.push({
        filePath: projectPath,
        error,
      });

      return {
        changed: false,
        problems,
      };
    }
  }

  private isSaving = false;

  async save(projectPathToSave?: string): Promise<void> {
    try {
      this.isSaving = true;

      const pagePathsToDelete = new Set(
        await this.fileAccess.glob(this.filePattern)
      );

      for (const [projectPath, project] of this.projects) {
        if (projectPathToSave && projectPath !== projectPathToSave) {
          continue;
        }

        const projectEmitter = new ProjectEmitter(project.project);
        const pages = projectEmitter.emit();

        for (const [pageName, pageNode] of pages) {
          const pagePath = path.join(projectPath, pageName + ".uimix");
          await this.fileAccess.writeFile(
            pagePath,
            Buffer.from(formatTypeScript(stringifyAsJSXFile(pageNode)))
          );
          pagePathsToDelete.delete(pagePath);
        }

        const usedImageHashes = usedImageHashesInProject(
          project.project.toJSON()
        );

        for (const [hash, image] of project.project.imageManager.images) {
          if (!usedImageHashes.has(hash)) {
            continue;
          }

          const decoded = dataUriToBuffer(image.url);
          const suffix = mime.extension(decoded.type) || "bin";
          await this.fileAccess.writeFile(
            path.join(projectPath, "src/images", `${hash}.${suffix}`),
            decoded
          );
        }

        // TODO: save manifest

        // const manifestPath = path.join(projectPath, this.uimixProjectFile);

        // let parsed: ProjectManifestJSON;
        // try {
        //   parsed = JSON.parse(
        //     (await this.fileAccess.readFile(manifestPath)).toString()
        //   ) as ProjectManifestJSON;
        // } catch (e) {
        //   parsed = {};
        // }
        // parsed.prebuiltAssets = project.manifest.prebuiltAssets;

        // TODO: avoid overwriting malformed uimix.json
        // await this.fileAccess.writeFile(
        //   manifestPath,
        //   Buffer.from(formatJSON(JSON.stringify(parsed)))
        // );

        project.pages = pages;
      }

      for (const pagePath of pagePathsToDelete) {
        await this.fileAccess.remove(pagePath);
      }
    } finally {
      this.isSaving = false;
    }
  }

  watch(onChange: () => void): () => void {
    console.log("start watching...");

    return this.fileAccess.watch(
      `{${this.filePattern},**/${this.projectBoundary}}`,
      async () => {
        try {
          if (this.isSaving) {
            return;
          }
          if ((await this.load()).changed) {
            onChange();
          }
        } catch (e) {
          console.error(e);
        }
      }
    );
  }
}
