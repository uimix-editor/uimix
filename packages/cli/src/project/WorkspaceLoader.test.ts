import { describe, it, expect, beforeEach, afterEach } from "vitest";
import { WorkspaceLoader } from "./WorkspaceLoader";
import * as fs from "fs";
import * as path from "path";
import shell from "shelljs";
import tmp from "tmp";
import { Project } from "@uimix/model/src/models/Project";
import { NodeFileAccess } from "./NodeFileAccess";

describe(WorkspaceLoader.name, () => {
  let tmpObj: tmp.DirResult;

  beforeEach(async () => {
    tmpObj = tmp.dirSync({
      unsafeCleanup: true,
    });
    shell.mkdir("-p", tmpObj.name + "/demo-project");
  });

  afterEach(async () => {
    tmpObj.removeCallback();
  });

  // it("loads", () => {
  //   const projectFiles = new ProjectFiles(tmpObj.name + "/demo-project");
  //   projectFiles.load();

  //   expect(projectFiles.toProjectJSON()).toMatchSnapshot();
  // });

  it("saves", async () => {
    const project = new Project();
    const page1 = project.pages.create("src/page1");
    const page2 = project.pages.create("src/page2");

    page1.node.append([project.nodes.create("frame", "frame1")]);
    page2.node.append([project.nodes.create("text", "text1")]);

    const innerProject = new Project();
    const innerPage1 = innerProject.pages.create("src/inner1");
    const innerPage2 = innerProject.pages.create("src/inner2");
    innerPage1.node.append([innerProject.nodes.create("frame", "inner1")]);
    innerPage2.node.append([innerProject.nodes.create("frame", "inner2")]);

    const deepInnerProject = new Project();
    deepInnerProject.pages.create("src/inner1");

    const innerProjectPath = tmpObj.name + "/demo-project/inner";
    const deepInnerProjectPath = tmpObj.name + "/demo-project/inner/inner";

    const loader = new WorkspaceLoader(
      new NodeFileAccess(tmpObj.name + "/demo-project")
    );
    loader.json = project.toJSON();
    loader.jsons.set(innerProjectPath, innerProject.toJSON());
    loader.jsons.set(deepInnerProjectPath, deepInnerProject.toJSON());
    await loader.save();

    fs.writeFileSync(tmpObj.name + "/demo-project/inner/package.json", "{}");
    fs.writeFileSync(
      tmpObj.name + "/demo-project/inner/inner/package.json",
      "{}"
    );

    expect(
      loader.projectPathForFile(tmpObj.name + "/demo-project/src/page1.uimix")
    ).toEqual(tmpObj.name + "/demo-project");
    expect(
      loader.projectPathForFile(
        tmpObj.name + "/demo-project/inner/src/inner1.uimix"
      )
    ).toEqual(tmpObj.name + "/demo-project/inner");

    const page1File = fs.readFileSync(
      tmpObj.name + "/demo-project/src/page1.uimix",
      "utf8"
    );
    expect(page1File).toMatchSnapshot();

    const page2File = fs.readFileSync(
      tmpObj.name + "/demo-project/src/page2.uimix",
      "utf8"
    );
    expect(page2File).toMatchSnapshot();

    const innerPage1File = fs.readFileSync(
      tmpObj.name + "/demo-project/inner/src/inner1.uimix",
      "utf8"
    );
    expect(innerPage1File).toMatchSnapshot();

    const projectFiles2 = await WorkspaceLoader.load(
      new NodeFileAccess(tmpObj.name + "/demo-project")
    );

    expect(projectFiles2.json).toEqual(project.toJSON());
    expect(projectFiles2.jsons.get(innerProjectPath)).toEqual(
      innerProject.toJSON()
    );
  });

  it("watches project", async () => {
    const project = new Project();
    const page1 = project.pages.create("src/page1");
    const page2 = project.pages.create("src/page2");

    page1.node.append([project.nodes.create("frame", "frame1")]);
    page2.node.append([project.nodes.create("text", "text1")]);

    const innerProject = new Project();
    const innerPage1 = innerProject.pages.create("src/inner1");
    innerPage1.node.append([innerProject.nodes.create("frame", "inner")]);

    const loader = new WorkspaceLoader(
      new NodeFileAccess(tmpObj.name + "/demo-project")
    );
    loader.json = project.toJSON();
    loader.jsons.set(
      path.resolve(loader.rootPath, "inner"),
      innerProject.toJSON()
    );
    await loader.save();

    fs.writeFileSync(tmpObj.name + "/demo-project/inner/package.json", "{}");

    let watchCount = 0;
    loader.watch(() => {
      watchCount++;
    });
    await new Promise((resolve) => setTimeout(resolve, 500));
    expect(watchCount).toBe(0);

    // saves not cause watch
    await loader.save();
    await new Promise((resolve) => setTimeout(resolve, 500));
    expect(watchCount).toBe(0);

    // change file
    fs.rmSync(tmpObj.name + "/demo-project/src/page1.uimix");
    await new Promise((resolve) => setTimeout(resolve, 500));
    expect(watchCount).toBe(1);

    // change inner file
    fs.rmSync(tmpObj.name + "/demo-project/inner/src/inner1.uimix");
    await new Promise((resolve) => setTimeout(resolve, 500));
    expect(watchCount).toBe(2);

    project.loadJSON(loader.json);

    expect(project.pages.all.map((page) => page.filePath)).toEqual([
      "src/page2",
    ]);
  });
});
