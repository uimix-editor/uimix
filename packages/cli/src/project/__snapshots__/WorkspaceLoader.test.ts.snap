// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WorkspaceLoader > saves 1`] = `
"export default (
  <page>
    <frame name=\\"\\" />
  </page>
);
"
`;

exports[`WorkspaceLoader > saves 2`] = `
"export default (
  <page>
    <text name=\\"\\" />
  </page>
);
"
`;

exports[`WorkspaceLoader > saves 3`] = `
"export default (
  <page>
    <frame name=\\"\\" />
  </page>
);
"
`;
